### PR TITLE
detect offline errors from chat fetch endpoints and have them set offline CORE-4833

### PR DESCRIPTION
**Pull request by mmaxim**
The patch does the following:

1.) Adds a function to detect errors that result from network problems. If an error like this is detected, then we set the `Offline` field on the response. This can happen if we think we are online, but lose that connectivity somewhere in the call. 
2.) Use the new error detector in chat endpoints that have some sort of offline functionality. Two exceptions are `DownloadAttachment` and `MarkThreadAsLocal`, which will just do nothing other than set `Offline` in this state.
3.) In `assertLoggedIn`, if we get an API network error, we just charge through and keep going.
4.) Change the superseder calculation for `PullLocalOnly` to use only messages we have cached. Otherwise, it delays the cached result from `GetThreadNonblock` if the thread has a superseded by message that we don't have the superseding message locally.

cc @chrisnojima

### CODE REVIEW

File: go/service/chat_local.go [hunk 1]
```diff
@@ -70,10 +70,52 @@ func (h *chatLocalHandler) getChatUI(sessionID int) libkb.ChatUI {
 	return h.BaseHandler.getChatUI(sessionID)
 }
 
+func (h *chatLocalHandler) isOfflineError(err error) bool {
+	// Check type
+	switch terr := err.(type) {
```
Comment (mmaxim) [inline go/service/chat_local.go]: Note: here is my list of errors that I think signify a connection problem.

File: go/service/chat_local.go [hunk 2]
```diff
@@ -70,10 +70,52 @@ func (h *chatLocalHandler) getChatUI(sessionID int) libkb.ChatUI {
...
+}
+
+func (h *chatLocalHandler) handleOfflineError(ctx context.Context, err error,
+	res chat1.OfflinableResult) error {
+
+	if h.isOfflineError(err) {
+		h.Debug(ctx, "handleOfflineError: setting offline: err: %s", err.Error())
+		res.SetOffline()
+
+		// Disconnect Gregor if we think we are online
```
Comment (patrickxb) [inline go/service/chat_local.go]: "Reconnect Gregor if we think we are offline"?  Is that what this is for?

Comment (mmaxim) [inline go/service/chat_local.go]: Yes

**Comment by mmaxim** [comment]
Oops I messed up this branch

**Comment by patrickxb** [comment]
looks fine, but jenkins is down.

**Pull request merged**